### ISSUE TITLE
bug: fullnameOverride for components does not work as expected

### ISSUE THREAD

**Issue by vhajdukd** [VictoriaMetrics/helm-charts#1716]
**Chart name and version**
chart: victoria-metrics-cluster
version: v0.14.6

**Describe the bug**
In recent versions the way names for components are generates was changed to incorporate for_each of every component available in the cluster chart defined in values.yaml.

I gotta say that the way you handle namings is now "PURE CANCER".. Suddenly changing service names can have huge impact on ingress, routing, gateways tons of components using in-cluster DNS routing.

Just take a look at this helper template:
```
{{- define "vm.fullname" -}}
  {{- include "vm.validate.args" . -}}
  {{- $Values := (.helm).Values | default .Values -}}
  {{- $Chart := (.helm).Chart | default .Chart -}}
  {{- $Release := (.helm).Release | default .Release -}}
  {{- $fullname := "" -}}
  {{- if .appKey -}}
    {{- $appKey := ternary (list .appKey) .appKey (kindIs "string" .appKey) -}}
    {{- $values := $Values -}}
    {{- $global := (index $Values.global $Chart.Name) | default dict -}}
    {{- range $ak := $appKey }}
      {{- if $values -}}
        {{- $values = (index $values $ak) | default dict -}}
      {{- end -}}
      {{- if $global -}}
        {{- $global = (index $global $ak) | default dict -}}
      {{- end -}}
      {{- if and (kindIs "map" $values) $values.name -}}
        {{- $fullname = $values.name -}}
      {{- else if and (kindIs "map" $values) $values.fullnameOverride -}}
        {{- $fullname = $values.fullnameOverride -}}
      {{- else if and (kindIs "map" $global) $global.name -}}
        {{- $fullname = $global.name -}}
      {{- end -}}
    {{- end }}
  {{- end -}}
  {{- if empty $fullname -}}
    {{- if $Values.fullnameOverride -}}
      {{- $fullname = $Values.fullnameOverride -}}
    {{- else if (dig $Chart.Name "fullnameOverride" "" ($Values.global)) -}}
      {{- $fullname = (dig $Chart.Name "fullnameOverride" "" ($Values.global)) -}}
    {{- else if ($Values.global).fullnameOverride -}}
      {{- $fullname = $Values.global.fullnameOverride -}}
    {{- else -}}
      {{- $name := default $Chart.Name $Values.nameOverride -}}
      {{- if contains $name $Release.Name -}}
        {{- $fullname = $Release.Name -}}
      {{- else -}}
        {{- $fullname = (printf "%s-%s" $Release.Name $name) }}
      {{- end -}}
    {{- end -}}
  {{- end -}}
  {{- with .prefix -}}
    {{- $fullname = printf "%s-%s" . $fullname -}}
  {{- end -}}
  {{- with .suffix -}}
    {{- $fullname = printf "%s-%s" $fullname . -}}
  {{- end -}}
  {{- $fullname | trunc 63 | trimSuffix "-" -}}
{{- end }}

{{- define "vm.plain.fullname" -}}
  {{- $suffix := .appKey -}}
  {{- $oldSuffix := .suffix -}}
  {{- if kindIs "slice" $suffix -}}
    {{- $suffix = last $suffix }}
  {{- end -}}
  {{- if $suffix -}}
    {{- with $oldSuffix -}}
      {{- $suffix = printf "%s-%s" $suffix . -}}
    {{- end -}}
    {{- $_ := set . "suffix" $suffix -}}
  {{- end -}}
  {{- include "vm.fullname" . -}}
  {{- $_ := set . "suffix" $oldSuffix -}}
{{- end -}}
```

How sane person is supposed to figure out how to retain the same service name between versions ? Especially when you cannot actively debug helm charts.

**Custom values**
N/A

Please reconsider this approach of changing naming strategies - 1) It's pure cancer reading this 2) its very error prone. Each and every component should have an option to define DEFINITIVE override for every component name.

`fullnameOverride` does not work, this helper template will always add additional suffix to the provided name no matter what you define.

Please add an option to override service names in case you want to modify component names in the future.

**Comment by AndrewChubatiuk** [comment]
hey @vhajdukd
okay, thanks for a feedback
have you tried the latest chart v0.14.9?
fullnameOverride should work properly there

**Comment by vhajdukd** [comment]
Yes I did try the latest version and issue still exist:
<img width="1121" alt="image" src="https://github.com/user-attachments/assets/d81e1b3d-de1f-4dde-a8f1-3f7821e77bd3">

Why is that suffix "-vminsert" even needed there if I have defined:
```
vminsert:
  fullnameOverride: vminsert-downsample-dest
```
This chart doesn't allow you to deploy multiple vminserts alongside each other.

**Comment by vhajdukd** [comment]
```
vminsert:
  enabled: true
  name: vminsert-downsample-dest
  fullnameOverride: vminsert-downsample-dest

  extraArgs:
    envflag.enable: "true"
    storageNode: vmstorage1:8400,vmstorage2:8400
    replicationFactor: "2"
    maxLabelsPerTimeseries: 80

  suppresStorageFQDNsRender: true
  automountServiceAccountToken: true

  horizontalPodAutoscaler:
    enabled: true
    maxReplicas: "5"
    minReplicas: "2"
    metrics:
      - type: Resource
        resource:
          name: cpu
          target:
            type: Utilization
            averageUtilization: 90

  probe:
    readiness:
      initialDelaySeconds: 5
      periodSeconds: 15
      timeoutSeconds: 5
      failureThreshold: 3
    liveness:
      initialDelaySeconds: 5
      periodSeconds: 15
      timeoutSeconds: 5
      failureThreshold: 3

  podDisruptionBudget:
    enabled: true
    minAvailable: 40%
    selector:
      matchLabels:
        app: vminsert-downsample-dest

  affinity:
    podAntiAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
        - weight: 100
          podAffinityTerm:
            labelSelector:
              matchLabels:
                app: vminsert-downsample-dest
            topologyKey: kubernetes.io/hostname
  topologySpreadConstraints:
    - maxSkew: 1
      topologyKey: topology.kubernetes.io/zone
      whenUnsatisfiable: ScheduleAnyway
      labelSelector:
        matchLabels:
          app: vminsert-downsample-dest

  resources:
    limits:
      cpu: 700m
      memory: 800Mi
    requests:
      cpu: 560m
      memory: 640Mi

  service:
    type: ClusterIP
    labels: {}
    servicePort: 80
    targetPort: 8480
    annotations:
      prometheus.io/scrape: 'true'
      prometheus.io/port: '8480'

  serviceMonitor:
    enabled: false
```

ps. In case I remove the `vminsert.name` then values of app labels are not retained.

**Comment by AndrewChubatiuk** [comment]
there was a problem while using both <component>.name and <component>.fullnameOverride also simplified template a bit. updates are available in 0.14.10

**Comment by vhajdukd** [comment]
Upgrade to common seems have helped. app label and name override are both retained:
<img width="1121" alt="image" src="https://github.com/user-attachments/assets/331bab9e-39b9-4ed3-877d-773a5c7dbe9e">
